include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1025 @@
+    const gradient_type
+    get_gradient(const unsigned int q) const;
+
+    /**
+     * Set entity index at which quadrature points are accessed. This can be,
+     * e.g., a cell index, a cell batch index, or a face batch index.
+     */
+    void
+    reinit(const unsigned int index);
+
+    /**
+     * Set cell and face_number at which quadrature points are accessed.
+     */
+    void
+    reinit(const unsigned int index_0, const unsigned int index_1);
+
+  private:
+    /**
+     * PrecomputedEvaluationDataView provides information where values are
+     * located.
+     */
+    const PrecomputedEvaluationDataView &view;
+
+    /**
+     * PrecomputedEvaluationData stores the actual values.
+     */
+    const PrecomputedEvaluationData<dim, n_components, value_type_> &data;
+
+    /**
+     * Offset to data after last call of `reinit()`.
+     */
+    unsigned int data_offset;
+  };
+
+} // namespace internal
+
+
+
+/**
+ * Communication objects know about the communication pattern.
+ * In case of (matrix-free) cells batches or faces batches
+ * a RemotePointEvaluation object stores the location of the
+ * remote points. @c batch_id_n_entities relates these points
+ * to the corresponding quadrature points of entity batches.
+ * For this the field stores batch IDs and the number of entities
+ * in the batch.
+ */
+template <int dim>
+struct FERemoteCommunicationObjectEntityBatches
+{
+  /**
+   * Object that is reinitialized with the remote points we want to access.
+   */
+  std::shared_ptr<Utilities::MPI::RemotePointEvaluation<dim>> rpe;
+
+  /**
+   * A vector that stores the batch IDs and the number of active entries in the
+   * batch related to the the remote points.
+   */
+  std::vector<std::pair<unsigned int, unsigned int>> batch_id_n_entities;
+
+  /**
+   * Function that gives access to @c batch_id_n_entities. The function is only
+   * used for a unified access to the vector that stores additional information
+   * on the points that are processed via @c rpe.
+   */
+  std::vector<std::pair<unsigned int, unsigned int>>
+  get_pntrs() const;
+};
+
+/**
+ * Similar as @c FERemoteCommunicationObjectEntityBatches.
+ * To relate the points from @c RemotePointEvaluation to
+ * quadrature points on corresponding cells, cell iterators
+ * have to be stored.
+ */
+template <int dim>
+struct FERemoteCommunicationObjectCells
+{
+  /**
+   * Object that is reinitialized with the remote points we want to access.
+   */
+  std::shared_ptr<Utilities::MPI::RemotePointEvaluation<dim>> rpe;
+
+  /**
+   * A vector that stores the cell iterators related to the the remote points.
+   */
+  std::vector<typename Triangulation<dim>::cell_iterator> cells;
+
+  /**
+   * Function that gives access to @c cells. The function is only used
+   * for a unified access to the vector that stores additional information on
+   * the points that are processed via @c rpe.
+   */
+  std::vector<typename Triangulation<dim>::cell_iterator>
+  get_pntrs() const;
```

Comment:
Not sure if the get_pntrs is a common function name we use, but if not, I would be more explicit with the function name

Reply:
I renamed to `get_communication_object_pntrs()` :)